### ISSUE TITLE
Failed  'CT normalize'

### ISSUE THREAD

**Issue by luanraguiars** [neurolabusc/Clinical#12]
Dear all,

When I tried to run the ‘CT normalize’ module of the Clinical Toolbox, I got the following error:

```
SPM present working directory:
	C:\Users\Neuroimager\Desktop\ANALISES_FINAIS

SPM12: spm_image (v7573)                           10:58:04 - 19/12/2023
========================================================================


------------------------------------------------------------------------
19-Dec-2023 10:58:52 - Running job #1
------------------------------------------------------------------------
19-Dec-2023 10:58:52 - Running 'CT normalize'
CT normalization version 7/7/2016
clinical_fix_ge_ct skipped: This image does not have unusual image intensities: C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\sub-02_ses1_space-orig_desc-CT.nii,1.
C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\sub-02_ses1_space-orig_desc-CT.nii center of brightness differs from current origin by -0x-19x9mm in X Y Z dimensions


------------------------------------------------------------------------
19-Dec-2023 10:58:53 - Running job #2
------------------------------------------------------------------------
19-Dec-2023 10:58:53 - Running 'Coregister: Estimate'

SPM12: spm_coreg (v7320)                           10:58:53 - 19/12/2023
========================================================================
Printing 'Graphics' figure to:
  C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\spm_2023Dec19.ps
Completed                               :          10:59:24 - 19/12/2023
19-Dec-2023 10:59:24 - Done    'Coregister: Estimate'
19-Dec-2023 10:59:24 - Done

C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\sub-02_ses1_space-orig_desc-CT.nii intensity range: 2967
 Ignore QFORM0 warning if reported next


------------------------------------------------------------------------
19-Dec-2023 10:59:27 - Running job #2
------------------------------------------------------------------------
19-Dec-2023 10:59:27 - Running 'Image Calculator'

SPM12: spm_imcalc (v6961)                          10:59:27 - 19/12/2023
========================================================================
ImCalc Image: C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\dxsub-02_ses1_space-orig_desc-CTlesion_mask.nii
19-Dec-2023 10:59:27 - Done    'Image Calculator'
19-Dec-2023 10:59:27 - Done

clinical_smoothmask: C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\xsub-02_ses1_space-orig_desc-CTlesion_mask.nii had 47043 voxels >0.001000
masking C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\sub-02_ses1_space-orig_desc-CT.nii with C:\Users\Neuroimager\Desktop\ANALISES_FINAIS\ssub-02_ses1_space-orig_desc-CTlesion_mask.nii using template C:\Users\Neuroimager\Desktop\Backup_LuanAguiar\Pacotes_Neuroimagem\spm12\toolbox\Clinical\scct.nii.
Conversion Normalise:EstWrite -> Old Normalise:EstWrite


------------------------------------------------------------------------
19-Dec-2023 10:59:28 - Running job #2
------------------------------------------------------------------------
19-Dec-2023 10:59:28 - Running 'Old Normalise: Estimate & Write'
Smoothing by 0 & 8mm..
Coarse Affine Registration..
Fine Affine Registration..
19-Dec-2023 10:59:29 - Failed  'Old Normalise: Estimate & Write'
Error using spm_affreg
Source and its weighting image must have same orientation.
In file "C:\Users\Neuroimager\Desktop\Backup_LuanAguiar\Pacotes_Neuroimagem\spm12\toolbox\OldNorm\spm_affreg.m" (v7377), function "spm_affreg" at line 66.
In file "C:\Users\Neuroimager\Desktop\Backup_LuanAguiar\Pacotes_Neuroimagem\spm12\toolbox\OldNorm\spm_normalise.m" (v5745), function "spm_normalise" at line 182.
In file "C:\Users\Neuroimager\Desktop\Backup_LuanAguiar\Pacotes_Neuroimagem\spm12\toolbox\OldNorm\spm_run_normalise.m" (v4873), function "spm_run_normalise" at line 33.

The following modules did not run:
Failed: Old Normalise: Estimate & Write

19-Dec-2023 10:59:29 - Failed  'CT normalize'
Error using MATLABbatch system
Job execution failed. The full log of this run can be found in MATLAB command window, starting with the lines (look for the line showing the exact #job as displayed in this error message)
------------------ 
Running job #2
------------------


The following modules did not run:
Failed: CT normalize
```
For more details:

![image](https://github.com/neurolabusc/Clinical/assets/128935795/a82d472f-eda7-4e19-83fd-9d0bbce9b0bf)
![image](https://github.com/neurolabusc/Clinical/assets/128935795/1479b653-3f52-40d2-a2b4-060eb7ec14b2)

Previously, I believed that it could be the step I performed before trying to perform normalization, which would be “reorientation” (defining the origin ‘anterior commissure’ and the orientation of the patient’s image manually in the SPM[Display]).

I would appreciate it very much if anyone could help me on this issue.

Thanks so much,

Luan.

**Comment by neurolabusc** [comment]
The error is here:
```
Source and its weighting image must have same orientation.
```
The image (`sub-02_ses1_space-orig_desc-CT.nii`) and its mask `sub-02_ses1_space-orig_desc-CTlesion_mask.nii`) must have identical dimensions and a qform. If the images have different physical orientations, you could try reslicing one to match the other by using [nii_reslice_target](https://github.com/rordenlab/spmScripts/blob/master/nii_reslice_target.m). If the sform/qform got corrupted in one, you could fix it with this Python script:

```
import nibabel as nib

def copy_affine_transform(source_path, target_path):
   # Load the source NIfTI image
   source_img = nib.load(source_path)

   # Extract the affine transform from the source image
   affine_transform = source_img.affine

   # Load the target NIfTI image
   target_img = nib.load(target_path)

   # Copy the affine transform to the target image
   target_img.set_qform(affine_transform)
   target_img.set_sform(affine_transform)

   # Save the modified target image
   nib.save(target_img, target_path)

if __name__ == "__main__":
   # Specify the paths to the source and target NIfTI images
   source_image_path = "./Source.nii.gz"
   target_image_path = "./Target.nii.gz"

   # Copy the affine transform from the source to the target image
   copy_affine_transform(source_image_path, target_image_path)

   print(f"Affine transform copied from {source_image_path} to {target_image_path}.")
```

**Comment by luanraguiars** [comment]
Hi Cris,

Thank you very much. I've tried both options, and neither has worked. 


The first option, the lesion mask is outside the lesion on the CT scan.

![Captura de tela de 2023-12-19 20-06-28](https://github.com/neurolabusc/Clinical/assets/128935795/6041e863-1034-47d7-a934-1817cfb21de4)

_______________________________________________

The second option (the script with the nibabel) reduces the nii file to kybtes (See photo below: 2, 'red'). 

![image](https://github.com/neurolabusc/Clinical/assets/128935795/bb470fdb-cc82-4ddb-a8b9-fb9d298ef17f)

In green (with number 1), I've printed the qform and sform of lesion.nii (first) and CT.nii (second).

_________________________________________________


Best,
Luan

**Comment by neurolabusc** [comment]
I can not provide further insights without seeing the data, feel free to send a link to a sample dataset [to my institutional email](https://sc.edu/study/colleges_schools/artsandsciences/psychology/our_people/directory/rorden_chris.php). if you want more assistance. Otherwise, I suggest you use matlab or `fslhd` to examine the headers of the image and mask to determine why they are not in register.

**Comment by luanraguiars** [comment]
Hi Rorden,

I ran the normalization on a Linux OS computer, and it worked. However, for some patients (who had lesion masks drawn on FLAIR), I had to orient the T1 (SPM12>Display) to be able to run MR segment-normalize.

Additionally, I used the "reslice" for a patient with CT segmentation because I noticed that the image and the lesion mask had different dimensions.

Thank you very much for the prompt response, especially during the holiday season. I am very grateful!

I wish you a belated Merry Christmas and an excellent 2024 with lots of health and peace.
